SimPEG/Regularization.py

```diff
@@ -61,320 +87,439 @@ def dim(self):
-        cell centered difference stencil (no cell lengths include) in the y-direction
+        cell centered difference stencil (no cell lengths include) in the
+        y-direction
+
         :rtype: scipy.sparse.csr_matrix
         :return: differencing matrix for active cells in the y-direction
         """
-        if self.dim < 2: return None
+        if self.dim < 2:
+            return None
         if getattr(self, '_cellDiffyStencil', None) is None:
 
-            self._cellDiffyStencil = self._Pafy.T * self.mesh._cellGradyStencil() * self._Pac
+            self._cellDiffyStencil = (
+                self.Pafy.T * self.mesh._cellGradyStencil() * self.Pac
+            )
         return self._cellDiffyStencil
 
     @property
     def cellDiffzStencil(self):
         """
-        cell centered difference stencil (no cell lengths include) in the y-direction
+        cell centered difference stencil (no cell lengths include) in the
+        y-direction
+
         :rtype: scipy.sparse.csr_matrix
         :return: differencing matrix for active cells in the y-direction
         """
-        if self.dim < 3: return None
+        if self.dim < 3:
+            return None
         if getattr(self, '_cellDiffzStencil', None) is None:
 
-            self._cellDiffzStencil = self._Pafz.T * self.mesh._cellGradzStencil() * self._Pac
+            self._cellDiffzStencil = (
+                self.Pafz.T * self.mesh._cellGradzStencil() * self.Pac
+            )
         return self._cellDiffzStencil
 
 
-class BaseRegularization(object):
+###############################################################################
+#                                                                             #
+#                          Single Regularization                              #
+#                                                                             #
+###############################################################################
+
+class BaseRegularization(ObjectiveFunction.L2ObjectiveFunction):
     """
-    **Base Regularization Class**
+    Base class for regularization. Inherit this for building your own
+    regularization. The base regularization assumes a weighted l2 style of
+    regularization. However, if you wish to employ a different norm, the
+    methods :meth:`_eval`, :meth:`deriv` and :meth:`deriv2` can be over-written
 
-    This is used to regularize the model space::
+    **Optional Inputs**
 
-        reg = Regularization(mesh)
+    :param BaseMesh mesh: SimPEG mesh
+    :param int nP: number of parameters
+    :param IdentityMap mapping: regularization mapping, takes the model from model space to the space you want to regularize in
+    :param numpy.ndarray mref: reference model
+    :param numpy.ndarray indActive: active cell indices for reducing the size
+    of differential operators in the definition of a regularization mesh
 
     """
 
     counter = None
+    mapPair = Maps.IdentityMap  #: Base class of expected maps
```

Comment:
`Props.Mapping` has a few things baked-in that are not strictly necessary here (eg. the reciprocal) so I think it is worth teasing that apart before having objective functions rely on it